compiler/rustc_session/src/errors.rs

```diff
@@ -32,6 +32,20 @@ pub struct FeatureDiagnosticForIssue {
     pub n: NonZeroU32,
 }
 
+#[derive(Subdiagnostic)]
+#[note(session_feature_suggest_upgrade_compiler)]
+pub struct SuggestUpgradeCompiler {
+    date: &'static str,
+}
+
+impl SuggestUpgradeCompiler {
+    pub fn new() -> Option<Self> {
+        let date = option_env!("CFG_VER_DATE")?;
+
+        Some(Self { date })
```

Comment:
This should check `sess.opts.unstable_options.ui_testing` and hardcode the result for those cases (to something like `YYYY-MM-DD` This will ensure that UI tests are not affected by the current day.

Reply:
Hi Nilstrieb, thank you for this feedback.

I have been trying to add a check for the `ui_testing` flag into the new diagnostic, however because it's stored in `config::Options`, it isn't available at the callsite, so I have been trying to propagate the flag down the callstack from the `Session`, however this has ballooned into a huge change touching many function signatures. I think this is doable, because the `ParseSess` comes from a `Session`, and so the opts can be accessed at that point, but I want to confirm this approach with you before moving forwards.

Reply:
I have completed this work and pushed a commit so that we can discuss this change. Please let me know if you think this is the right approach.

The new message is: `this compiler was built on YYYY-MM-DD; consider upgrading if it is out of date` when ui testing

Reply:
I accidentally created a merge commit and had to do some magic to fix the history. The commit to review is 1078e09d5bdb4fa575fc9f58092f862e66cb03c9

Reply:
Hm, I don't think having to pass it is good. I would prefer changing the first argument to be a `Sess` and have an extra function for those callers that only have a `ParseSess` (and don't bother emitting the suggestion for those).

Reply:
I agree it would be nicer to pass the `Session` down instead. I see what you mean about having a separate function (or maybe an `Option<bool>` parameter?), however I'm not sure if the cases we need to cover already have access to the `Session` or not (and I'm not sure what cases those would be). Perhaps I could instead promote `ParseSess` to `Session` in the call hierarchy..? Hmm

Reply:
Most callers should have access to the session. Maybe you need to promote a few extra arguments like that, yeah.